### PR TITLE
Implement models for 1.4 spec, etc...

### PR DESCRIPTION
**Pull request by DarthHater**
THE OMNIBUS!

So far:
- vulnerabilities
- additional changes such as 
  - Signature (JSON only), and adding the rest of the 1.4 specific changes
  - ReleaseNotes
  - Adds Signature and ReleaseNotes to Service
  - Adds ExternalReference to Tool

Passes the test suite that @stevespringett implemented, with all the valid and invalid examples

To go:
- @mr-zepol is adding parser specific tests and generator specific tests to check that the Object Model is populated correctly, per expectations!

### CODE REVIEW

File: src/main/java/org/cyclonedx/model/ReleaseNotes.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+@JsonPropertyOrder({
+    "type",
+    "title",
+    "featuredImage",
+    "socialImage",
+    "description",
+    "timestamp",
+    "aliases",
+    "tags",
+    "resolves",
```
Comment (madpah) [inline src/main/java/org/cyclonedx/model/ReleaseNotes.java:41]: Missing `notes`?

**Review by madpah — CHANGES_REQUESTED** [review]
All looks good @DarthHater - one minor query.

File: src/main/java/org/cyclonedx/model/ReleaseNotes.java (hunk 1 above)
Comment (DarthHater) [reply src/main/java/org/cyclonedx/model/ReleaseNotes.java:41]: In: 4b971af and THANKS!

File: src/main/java/org/cyclonedx/model/Signature.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    "signers",
+    "chain",
+    "algorithm",
+    "keyId",
+    "publicKey",
+    "certificatePath",
+    "excludes",
+    "value"
+})
+public class Signature
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply src/main/java/org/cyclonedx/model/Signature.java:42]: In: 66d3afa

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    "analysis",
+    "affects"
+})
+public class Vulnerability
+{
+  public Vulnerability() {}
+
+  @JacksonXmlProperty(isAttribute = true, localName = "bom-ref")
+  @JsonProperty("bom-ref")
+  private String bomRef;
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java:64]: In: 66d3afa

File: src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    "advisories",
+    "created",
+    "published",
+    "updated",
+    "credits",
+    "tools",
+    "analysis",
+    "affects"
+})
+public class Vulnerability
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply src/main/java/org/cyclonedx/model/vulnerability/Vulnerability.java:58]: In: 66d3afa

**Comment by DarthHater** [comment]
> I wonder if we should update the example pom in the README to point to 6.0.0? Currently it says:
> 
> ```
> <dependency>
>     <groupId>org.cyclonedx</groupId>
>     <artifactId>cyclonedx-core-java</artifactId>
>     <version>5.0.4</version>
> </dependency>
> ```

In: f0402ae

File: src/main/java/org/cyclonedx/model/Service.java [hunk 5]
```diff
@@ -221,4 +227,12 @@ public void addProperty(Property property) {
     public void setServices(List<Service> services) {
         this.services = services;
     }
+
+    public ReleaseNotes getReleaseNotes() { return releaseNotes; }
+
+    public void setReleaseNotes() { this.releaseNotes = releaseNotes; }
```
[the remark this answers is not in the document]
Comment (DarthHater) [reply src/main/java/org/cyclonedx/model/Service.java]: In: fd69fa7

File: src/main/java/org/cyclonedx/model/Evidence.java [hunk 6]
```diff
@@ -35,6 +36,7 @@
 public class Evidence extends ExtensibleElement {
 
     private LicenseChoice license;
+    @JacksonXmlElementWrapper(localName = "copyright")
```
Comment (DarthHater) [inline src/main/java/org/cyclonedx/model/Evidence.java:39]: Try `    @JacksonXmlElementWrapper(useWrapping = false)`, I think that's what you are after

**Comment by stevespringett** [comment]
Tested with DT 4.4 and seems to work fine. No breaking changes that I can see. DT doesn't support the entire CDX 1.4 object model so unable to test that yet. But it looks good to me thus far.

**Comment by mealingr** [comment]
Should "1.4" also be added here?
https://github.com/CycloneDX/cyclonedx-core-java/blob/b5b28e813cca82796cb9d8d31cae1d13ee19687c/src/main/java/org/cyclonedx/model/Pedigree.java#L48
https://cyclonedx.org/docs/1.4/json/#components_items_pedigree_patches

and here?
https://github.com/CycloneDX/cyclonedx-core-java/blob/b5b28e813cca82796cb9d8d31cae1d13ee19687c/src/main/java/org/cyclonedx/model/Service.java#L70
https://cyclonedx.org/docs/1.4/json/#services_items_properties

File: src/main/java/org/cyclonedx/model/ExternalReference.java [hunk 7]
```diff
@@ -66,6 +66,8 @@
         BUILD_META("build-meta"),
         @JsonProperty("build-system")
         BUILD_SYSTEM("build-system"),
+        @JsonProperty("release_notes")
+        RELEASE_NOTES("release_notes"),
```
Comment (mealingr) [inline src/main/java/org/cyclonedx/model/ExternalReference.java:70]:
```suggestion
        @JsonProperty("release-notes")
        RELEASE_NOTES("release-notes"),
```
https://github.com/CycloneDX/specification/blob/1.4/schema/bom-1.4.xsd#L776
https://github.com/CycloneDX/specification/blob/1.4/schema/bom-1.4.schema.json#L882

File: src/main/java/org/cyclonedx/model/Signature.java [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+    private Kty kty;
+
+    public enum Crv {
+      @JsonProperty("P-256")
+      P_256("P-256"),
+      @JsonProperty("P-384")
+      P_384("P-384"),
+      @JsonProperty("P-521")
+      P_521("P-521");
```
Comment (mealingr) [inline src/main/java/org/cyclonedx/model/Signature.java:199]: I think there might be these other two as well
```suggestion
      P_521("P-521"),
      @JsonProperty("Ed25519")
      Ed25519("Ed25519"),
      @JsonProperty("Ed448")
      Ed448("Ed448");
```
https://cyclonedx.org/docs/1.4/json/#signature_oneOf_i2_publicKey_allOf_i1_then_crv

**Comment by mealingr** [comment]
I think I spotted a few unused imports like
https://github.com/CycloneDX/cyclonedx-core-java/blob/b5b28e813cca82796cb9d8d31cae1d13ee19687c/src/main/java/org/cyclonedx/model/vulnerability/Vulnerability10.java#L33
may be good to go through the changed files and check for others

**Pull request closed without merging**